### PR TITLE
Allow specifying python version relative to user's home directory.

### PR DESCRIPTION
**Pull request by meunierd**
The main use case for this PR is to more easily work with user-local python installations, particularly pyenv's. This way a user can specify `~/.pyenv/versions/3.4.3/bin/python`. This is particularly useful when working with projects across many python versions, where it doesn't make sense to have them all available on your PATH, while still keeping the .pre-commit-config.yaml portable by not specifying an absolute path.

### CODE REVIEW

**Comment by asottile** [comment]
Looks good, mind writing a test so this doesn't regress?

**Comment by meunierd** [comment]
Yeah! I just wasn't sure where it should live in the test suite. Any suggestions?

**Comment by asottile** [comment]
Probably this file would suffice: https://github.com/pre-commit/pre-commit/blob/master/tests/repository_test.py

Something similar to what test_switch_language_versions_doesnt_clobber does but set up a virtualenv in the home directory

**Comment by meunierd** [comment]
Updated with a test. The intention wasn't to specify the target of the virtualenv itself but rather the python binary passed to with `-p`

File: tests/repository_test.py [hunk 1]
```diff
@@ -414,3 +416,12 @@ def test_local_repository():
...
     assert len(local_repo.hooks) == 1
+
+
+def test_norm_version_expanduser():
+    old_home = os.environ.get('HOME', None)
+    test_home = '/Users/someuser'
+    os.environ['HOME'] = test_home
+    expected_path = '/Users/someuser/.pyenv/versions/3.4.3/bin/python'
+    assert norm_version('~/.pyenv/versions/3.4.3/bin/python') == expected_path
+    os.environ['HOME'] = old_home
```
Comment (asottile) [inline tests/repository_test.py]: you'll want to use:

```python

with mock.patch.dict(os.environ, {'HOME': test_home}):
    # run your assertion
```

right now when the assertion fails, the environment is polluted for the rest of the test.

I think you might also need to twiddle the file paths so the test passes on windows.

**Comment by meunierd** [comment]
Finally updated this PR. Sorry about the delay/noise. Windows is a pain.

File: pre_commit/languages/python.py [hunk 2]
```diff
@@ -31,11 +31,13 @@ def in_env(repo_cmd_runner, language_version):
 
 
 def norm_version(version):
+    version = os.path.expanduser(version)
     if os.name == 'nt':  # pragma: no cover (windows)
         if not distutils.spawn.find_executable(version):
             # The default place for python on windows is:
             # C:\PythonXX\python.exe
             version = r'C:\{0}\python.exe'.format(version.replace('.', ''))
+            version = version.replace(r'C:\\', 'C:\\')
```
Comment (asottile) [inline pre_commit/languages/python.py]: hmm, why was this necessary?

Maybe move the expanduser after the windows branch?

File: tests/repository_test.py [hunk 3]
```diff
@@ -414,3 +416,15 @@ def test_local_repository():
     with pytest.raises(NotImplementedError):
         local_repo.manifest
     assert len(local_repo.hooks) == 1
+
+
+def test_norm_version_expanduser():
+    home = os.path.expanduser('~')
+    if os.name == 'nt':
```
Comment (asottile) [inline tests/repository_test.py]: `# pragma: no cover` these two branches so coverage doesn't complain :)

**Comment by asottile** [comment]
Sweet, looks good -- minor nitpicks left :D

File: pre_commit/languages/python.py (hunk 2 above)
Comment (meunierd) [inline pre_commit/languages/python.py]: The expansion of `~` on Windows has a leading slash for some reason.

**Comment by meunierd** [comment]
Nit picked. I'm not sure of a better approach for the C:\\ replace though.

File: pre_commit/languages/python.py (hunk 2 above)
Comment (asottile) [inline pre_commit/languages/python.py]: Ah, maybe remove the C:\ part entirely from the format string?  iirc \foo\bar is the same as C:\foo\bar anyway and then pre-commit cares less about what your default drive is

**Comment by meunierd** [comment]
Updated with a more explicit handling of the leading slash issue.

**Comment by asottile** [comment]
Sweet! Thanks for the PR, be sure to add yourself on https://github.com/pre-commit/pre-commit.github.io

**Comment by meunierd** [comment]
Thanks! Will do.
